src/main/groovy/wooga/gradle/compat/ProjectLayout.groovy

```diff
@@ -0,0 +1,37 @@
+package wooga.gradle.compat
+
+import org.gradle.api.Project
+import org.gradle.api.file.Directory
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.file.RegularFile
+import org.gradle.api.file.RegularFileProperty
+import org.gradle.api.provider.Provider
+
+class ProjectLayout {
+
+    final Project project
+    final String gradleVersion
+    final Object compatFactory
+
+    ProjectLayout(Project project) {
+        this.project = project
+        gradleVersion = project.gradle.gradleVersion
+        compatFactory = gradleVersion.startsWith("6.") ? project.objects : project.layout
+    }
+
+    DirectoryProperty directoryProperty() {
+        (DirectoryProperty) compatFactory.invokeMethod("directoryProperty", null)
+    }
+
+    DirectoryProperty directoryProperty(Provider<? extends Directory> var1) {
+        (DirectoryProperty) compatFactory.invokeMethod("directoryProperty", var1)
+    }
+
+    RegularFileProperty fileProperty() {
+        (RegularFileProperty) compatFactory.invokeMethod("fileProperty", null)
+    }
+
+    RegularFileProperty fileProperty(Provider<? extends RegularFile> var1) {
+        (RegularFileProperty) compatFactory.invokeMethod("fileProperty", var1)
+    }
+}
```

Comment:
Why is this needed?

Reply:
One of my major issues with gradle is how they create and stabalize APIs.
The provider API was `@incubating` in gradle 4 and 5. I didn't check in gradle 6 but what they did is to move two property factory methods. Since gradle 6 is not API compatible we have to somehow make sure to keep the compatibility. Because we use gradle 4.10.2 for our projects and now also gradle 6.4 for android we need to stay compatible across all versions.

The methods `directoryProperty` and `fileProperty` used to be located in `ProjectLayout` instance. I honestly never liked this because you had to remember that these two properties had to be constructed from a different factory. So the gradle developers thought so too and moved these two methods into the `ObjectFactory` instance with gradle 6. I created this little helper class to call at runtime the correct location for the methods. When we move to gradle 6 for good we can remove this construct. I might move this out into a helper lib if I see the need in another plugin later.